docs/man/man3/ReuseDistance.3
.TH "ReuseDistance" 3 "21 Oct 2012" "Version 0.01" "ReuseDistance" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ReuseDistance \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <ReuseDistance.hpp>\fP
.PP
Inherited by \fBSpatialLocality\fP.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBReuseDistance\fP (uint64_t w, uint64_t b)"
.br
.ti -1c
.RI "\fBReuseDistance\fP (uint64_t w)"
.br
.ti -1c
.RI "virtual \fB~ReuseDistance\fP ()"
.br
.ti -1c
.RI "virtual void \fBPrint\fP (std::ostream &f, bool annotate=false)"
.br
.ti -1c
.RI "virtual void \fBPrint\fP (bool annotate=false)"
.br
.ti -1c
.RI "void \fBPrintFormat\fP (std::ostream &f)"
.br
.ti -1c
.RI "virtual void \fBProcess\fP (\fBReuseEntry\fP &addr)"
.br
.ti -1c
.RI "void \fBProcess\fP (\fBReuseEntry\fP *addrs, uint64_t count)"
.br
.ti -1c
.RI "void \fBProcess\fP (std::vector< \fBReuseEntry\fP > rs)"
.br
.ti -1c
.RI "void \fBProcess\fP (std::vector< \fBReuseEntry\fP * > addrs)"
.br
.ti -1c
.RI "\fBReuseStats\fP * \fBGetStats\fP (uint64_t id)"
.br
.ti -1c
.RI "void \fBGetIndices\fP (std::vector< uint64_t > &ids)"
.br
.ti -1c
.RI "virtual void \fBGetActiveAddresses\fP (std::vector< uint64_t > &addrs)"
.br
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const uint64_t \fBDefaultBinIndividual\fP = 32"
.br
.ti -1c
.RI "static const uint64_t \fBInfinity\fP = INFINITY_REUSE"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "void \fBInit\fP (uint64_t w, uint64_t b)"
.br
.ti -1c
.RI "virtual \fBReuseStats\fP * \fBGetStats\fP (uint64_t id, bool gen)"
.br
.ti -1c
.RI "virtual const std::string \fBDescribe\fP ()"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "reuse_map_type< uint64_t, \fBReuseStats\fP * > \fBstats\fP"
.br
.ti -1c
.RI "uint64_t \fBcapacity\fP"
.br
.ti -1c
.RI "uint64_t \fBsequence\fP"
.br
.ti -1c
.RI "uint64_t \fBbinindividual\fP"
.br
.ti -1c
.RI "uint64_t \fBmaxtracking\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Tracks reuse distances for a memory address stream. Keep track of the addresses within a specific window of history, whose size can be finite or infinite. For basic usage, see the documentation at http://bit.ly/ScqZVj for the constructors, the Process methods and the Print methods. Also see the simple test file test/test.cpp included in this source package. 
.PP
Definition at line 86 of file ReuseDistance.hpp.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "ReuseDistance::ReuseDistance (uint64_t w, uint64_t b)"Contructs a \fBReuseDistance\fP object.
.PP
\fBParameters:\fP
.RS 4
\fIw\fP The maximum window size, or alternatively the maximum possible reuse distance that this tool will find. No window/distance limit is imposed if \fBReuseDistance::Infinity\fP is used, though you could easily run of of memory. 
.br
\fIb\fP All distances not greater than b will be tracked individually. All distances are tracked individually if b == \fBReuseDistance::Infinity\fP. Beyond individual tracking, distances are tracked in bins whose boundaries are the powers of two greater than b (and not exeeding w, of course). 
.RE
.PP

.PP
Definition at line 56 of file ReuseDistance.cpp.
.SS "ReuseDistance::ReuseDistance (uint64_t w)"Contructs a \fBReuseDistance\fP object. Equivalent to calling the other constructor with b == \fBReuseDistance::DefaultBinIndividual\fP 
.PP
Definition at line 60 of file ReuseDistance.cpp.
.SS "ReuseDistance::~ReuseDistance ()\fC [virtual]\fP"Destroys a \fBReuseDistance\fP object. 
.PP
Definition at line 64 of file ReuseDistance.cpp.
.SH "Member Function Documentation"
.PP 
.SS "virtual const std::string ReuseDistance::Describe ()\fC [inline, protected, virtual]\fP"
.PP
Definition at line 108 of file ReuseDistance.hpp.
.SS "void ReuseDistance::GetActiveAddresses (std::vector< uint64_t > & addrs)\fC [virtual]\fP"Get a std::vector containing all of the addresses currently in this \fBReuseDistance\fP object's active window.
.PP
\fBParameters:\fP
.RS 4
\fIaddrs\fP A std::vector which will contain the addresses. It is an error to pass this vector non-empty (that is addrs.size() == 0 is enforced at runtime).
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
Reimplemented in \fBSpatialLocality\fP.
.PP
Definition at line 90 of file ReuseDistance.cpp.
.SS "void ReuseDistance::GetIndices (std::vector< uint64_t > & ids)"Get a std::vector containing all of the unique indices processed by this \fBReuseDistance\fP object.
.PP
\fBParameters:\fP
.RS 4
\fIids\fP A std::vector which will contain the ids. It is an error to pass this vector non-empty (that is addrs.size() == 0 is enforced at runtime).
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
Definition at line 82 of file ReuseDistance.cpp.
.SS "\fBReuseStats\fP * ReuseDistance::GetStats (uint64_t id)"Get the \fBReuseStats\fP object associated with some unique id.
.PP
\fBParameters:\fP
.RS 4
\fIid\fP The unique id.
.RE
.PP
\fBReturns:\fP
.RS 4
The \fBReuseStats\fP object associated with parameter id, or NULL if no \fBReuseStats\fP is associate with id. 
.RE
.PP

.PP
Definition at line 286 of file ReuseDistance.cpp.
.SS "\fBReuseStats\fP * ReuseDistance::GetStats (uint64_t id, bool gen)\fC [protected, virtual]\fP"
.PP
Definition at line 245 of file ReuseDistance.cpp.
.SS "void ReuseDistance::Init (uint64_t w, uint64_t b)\fC [protected]\fP"
.PP
Definition at line 40 of file ReuseDistance.cpp.
.SS "void ReuseDistance::Print (bool annotate = \fCfalse\fP)\fC [virtual]\fP"Print statistics for this \fBReuseDistance\fP to std::cout. See the other version of \fBReuseDistance::Print\fP for information about output format.
.PP
\fBParameters:\fP
.RS 4
\fIannotate\fP Also print annotations describing the meaning of output fields, preceded by a '#'.
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
Definition at line 100 of file ReuseDistance.cpp.
.SS "virtual void ReuseDistance::Print (std::ostream & f, bool annotate = \fCfalse\fP)\fC [virtual]\fP"Print statistics for this \fBReuseDistance\fP to an output stream. The first line of the output is 7 tokens: [1] a string identifier for the class (REUSESTATS or SPATIALSTATS), [2] the capacity or window size (0 == unlimited), [3] the maximum individual value being tracked, above which values are tracked by bins whose boundaries are powers of 2, [4] the maximum value to track, above which any value is considered a miss. For \fBReuseDistance\fP, this is equal to the capacity, for subclasses this can be different. [6] the number of ids that will be printed, [6] the total number of accesses made (the number of \fBReuseEntry\fP elements that were Process'ed) and [7] the number of accesses that cold-misses or were outside the window range. The stats for individual ids are printed on subsequent lines. The printing of each id begins with a line which is comprised of 4 tokens: [1] a string identifier (REUSEID or SPATIALID), [2] the id, [3] the number of accesses to that id and [4] the number of accesses for that id that were cold-misses or were outside the window range. Each subsequent line contains information about a single bin for that id. These lines have 3 tokens: [1] and [2] the lower and upper boundaries (both inclusive) of the bin and [3] the number of accesses falling into that bin. See also \fBReuseDistance::PrintFormat\fP
.PP
\fBParameters:\fP
.RS 4
\fIf\fP The output stream to print results to. 
.br
\fIannotate\fP Also print annotations describing the meaning of output fields, preceded by a '#'.
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void ReuseDistance::PrintFormat (std::ostream & f)"Print information about the output format of \fBReuseDistance\fP or one of its subclasses
.PP
\fBParameters:\fP
.RS 4
\fIf\fP The stream to receive the output.
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void ReuseDistance::Process (std::vector< \fBReuseEntry\fP * > addrs)"Process multiple memory addresses. Equivalent to calling Process on each element of the input vector.
.PP
\fBParameters:\fP
.RS 4
\fIaddrs\fP A std::vector of memory addresses to process.
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void ReuseDistance::Process (std::vector< \fBReuseEntry\fP > rs)"Process multiple memory addresses. Equivalent to calling Process on each element of the input vector.
.PP
\fBParameters:\fP
.RS 4
\fIaddrs\fP A std::vector of memory addresses to process.
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.SS "void ReuseDistance::Process (\fBReuseEntry\fP * addrs, uint64_t count)"Process multiple memory addresses. Equivalent to calling Process on each element of the input array.
.PP
\fBParameters:\fP
.RS 4
\fIaddrs\fP An array of structures describing memory addresses to process. 
.br
\fIcount\fP The number of elements in addrs.
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
Definition at line 104 of file ReuseDistance.cpp.
.SS "void ReuseDistance::Process (\fBReuseEntry\fP & addr)\fC [virtual]\fP"Process a single memory address.
.PP
\fBParameters:\fP
.RS 4
\fIaddr\fP The structure describing the memory address to process.
.RE
.PP
\fBReturns:\fP
.RS 4
none 
.RE
.PP

.PP
Reimplemented in \fBSpatialLocality\fP.
.PP
Definition at line 124 of file ReuseDistance.cpp.
.SH "Member Data Documentation"
.PP 
.SS "uint64_t \fBReuseDistance::binindividual\fP\fC [protected]\fP"
.PP
Definition at line 103 of file ReuseDistance.hpp.
.SS "uint64_t \fBReuseDistance::capacity\fP\fC [protected]\fP"
.PP
Definition at line 101 of file ReuseDistance.hpp.
.SS "const uint64_t \fBReuseDistance::DefaultBinIndividual\fP = 32\fC [static]\fP"
.PP
Definition at line 112 of file ReuseDistance.hpp.
.SS "const uint64_t \fBReuseDistance::Infinity\fP = INFINITY_REUSE\fC [static]\fP"
.PP
Definition at line 113 of file ReuseDistance.hpp.
.SS "uint64_t \fBReuseDistance::maxtracking\fP\fC [protected]\fP"
.PP
Definition at line 104 of file ReuseDistance.hpp.
.SS "uint64_t \fBReuseDistance::sequence\fP\fC [protected]\fP"
.PP
Definition at line 102 of file ReuseDistance.hpp.
.SS "reuse_map_type<uint64_t, \fBReuseStats\fP*> \fBReuseDistance::stats\fP\fC [protected]\fP"
.PP
Definition at line 99 of file ReuseDistance.hpp.

.SH "Author"
.PP 
Generated automatically by Doxygen for ReuseDistance from the source code.


ReuseDistance.cpp
/* 
 * This file is part of the ReuseDistance tool.
 * 
 * Copyright (c) 2012, University of California Regents
 * All rights reserved.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <ReuseDistance.hpp>

using namespace std;

#define REUSE_DEBUG
#ifdef REUSE_DEBUG
#define debug_assert(...) assert(__VA_ARGS__)
#else
#define debug_assert(...)
#endif

inline uint64_t uint64abs(uint64_t a){
    if (a < 0x8000000000000000L){
        return a;
    } else {
        return 0 - a;
    }
}

void ReuseDistance::Init(uint64_t w, uint64_t b){
    capacity = w;
    binindividual = b;
    maxtracking = capacity;

    current = 0;
    sequence = 1;

    window = newtree234();
    assert(window);

    mwindow.clear();

    assert(ReuseDistance::Infinity == NULL && "NULL is non-zero!?");
}

ReuseDistance::ReuseDistance(uint64_t w, uint64_t b){
    ReuseDistance::Init(w, b);
}

ReuseDistance::ReuseDistance(uint64_t w){
    ReuseDistance::Init(w, DefaultBinIndividual);
}

ReuseDistance::~ReuseDistance(){
    for (reuse_map_type<uint64_t, ReuseStats*>::const_iterator it = stats.begin(); it != stats.end(); it++){
        uint64_t id = it->first;
        delete stats[id];
    }

    debug_assert(current == count234(window));
    while (current){
        delete (ReuseEntry*)delpos234(window, 0);
        current--;
    }
    freetree234(window);
}

uint64_t ReuseStats::GetMissCount(){
    return distcounts[invalid];
}

void ReuseDistance::GetIndices(std::vector<uint64_t>& ids){
    assert(ids.size() == 0);
    for (reuse_map_type<uint64_t, ReuseStats*>::const_iterator it = stats.begin(); it != stats.end(); it++){
        uint64_t id = it->first;
        ids.push_back(id);
    }
}

void ReuseDistance::GetActiveAddresses(std::vector<uint64_t>& addrs){
    assert(addrs.size() == 0);
    debug_assert(current == count234(window));

    for (int i = 0; i < current; i++){
        ReuseEntry* r = (ReuseEntry*)index234(window, i);
        addrs.push_back(r->address);
    }
}

void ReuseDistance::Print(bool annotate){
    Print(cout, annotate);
}

void ReuseDistance::Process(ReuseEntry* rs, uint64_t count){
    for (uint32_t i = 0; i < count; i++){
        Process(rs[i]);
    }
}

void ReuseDistance::Process(vector<ReuseEntry> rs){
    for (vector<ReuseEntry>::const_iterator it = rs.begin(); it != rs.end(); it++){
        ReuseEntry r = *it;
        Process(r);
    }
}

void ReuseDistance::Process(vector<ReuseEntry*> rs){
    for (vector<ReuseEntry*>::const_iterator it = rs.begin(); it != rs.end(); it++){
        ReuseEntry* r = *it;
        Process((*r));
    }
}

void ReuseDistance::Process(ReuseEntry& r){
    uint64_t addr = r.address;
    uint64_t id = r.id;
    uint64_t mres = mwindow.count(addr);

    ReuseStats* stats = GetStats(id, true);

    int dist = 0;
    ReuseEntry* result;
    if (mres){
        mres = mwindow[addr];
        ReuseEntry key;
        key.address = addr;
        key.__seq = mres;

        result = findrelpos234(window, &key, &dist);
        debug_assert(result);

        if (capacity != ReuseDistance::Infinity){
            debug_assert(current - dist <= capacity);
        }
        stats->Update(current - dist);
    } else {
        stats->Update(ReuseDistance::Infinity);
    }

    // recycle a slot when possible
    ReuseEntry* slot = NULL;
    if (mres || (capacity != ReuseDistance::Infinity && current >= capacity)){
        slot = (ReuseEntry*)delpos234(window, dist);
        debug_assert(mwindow[slot->address]);
        mwindow.erase(slot->address);
        debug_assert(count234(window) == mwindow.size());
    } else {
        slot = new ReuseEntry();
        current++;
    }
    
    mwindow[addr] = sequence;

    slot->__seq = sequence;
    slot->address = addr;
    add234(window, slot);

    debug_assert(count234(window) == mwindow.size());
    debug_assert(mwindow.size() <= current);

    sequence++;
}

void ReuseDistance::PrintFormat(ostream& f){
    f << "# "
      << Describe() << "STATS"
      << TAB << "<window_size>"
      << TAB << "<bin_indiv>"
      << TAB << "<max_track>"
      << TAB << "<id_count>"
      << TAB << "<tot_access>"
      << TAB << "<tot_miss>"
      << ENDL;

    f << "# "
      << TAB << Describe() << "ID"
      << TAB << "<id>"
      << TAB << "<id_access>"
      << TAB << "<id_miss>"
      << ENDL;
}

void ReuseStats::PrintFormat(ostream& f){
    f << "# "
      << TAB 
      << TAB << "<bin_lower_bound>"
      << TAB << "<bin_upper_bound>"
      << TAB << "<bin_count>"
      << ENDL;
}

void ReuseDistance::Print(ostream& f, bool annotate){
    vector<uint64_t> keys;
    for (reuse_map_type<uint64_t, ReuseStats*>::const_iterator it = stats.begin(); it != stats.end(); it++){
        keys.push_back(it->first);
    }
    sort(keys.begin(), keys.end());

    uint64_t tot = 0, mis = 0;
    for (vector<uint64_t>::const_iterator it = keys.begin(); it != keys.end(); it++){
        uint64_t id = (*it);
        ReuseStats* r = (ReuseStats*)stats[id];
        tot += r->GetAccessCount();
        mis += r->GetMissCount();
    }

    if (annotate){
        ReuseDistance::PrintFormat(f);
        ReuseStats::PrintFormat(f);
    }

    f << Describe() << "STATS"
      << TAB << dec << capacity
      << TAB << binindividual
      << TAB << maxtracking
      << TAB << keys.size()
      << TAB << tot
      << TAB << mis
      << ENDL;

    for (vector<uint64_t>::const_iterator it = keys.begin(); it != keys.end(); it++){
        uint64_t id = (*it);
        ReuseStats* r = (ReuseStats*)stats[id];

        f << TAB << Describe() << "ID"
          << TAB << dec << id
          << TAB << r->GetAccessCount()
          << TAB << r->GetMissCount()
          << ENDL;

        r->Print(f);
    }
}

ReuseStats* ReuseDistance::GetStats(uint64_t id, bool gen){
    ReuseStats* s = stats[id];
    if (s == NULL && gen){
        s = new ReuseStats(id, binindividual, capacity, ReuseDistance::Infinity);
        stats[id] = s;
    }
    return s;
}

// this should be fast as possible. This code is from http://graphics.stanford.edu/~seander/bithacks.html#IntegerLog
static const uint64_t b[] = {0x2L, 0xCL, 0xF0L, 0xFF00L, 0xFFFF0000L, 0xFFFFFFFF00000000L};
static const uint32_t S[] = {1, 2, 4, 8, 16, 32};
inline uint64_t ShaveBitsPwr2(uint64_t val){
    val -= 1;
    register uint64_t r = 0; // result of log2(v) will go here
    for (int32_t i = 5; i >= 0; i--){
        if (val & b[i]){
            val = val >> S[i];
            r |= S[i];
        }
    }
    return (2 << r);
}

uint64_t ReuseStats::GetBin(uint64_t value){
    // not a valid value
    if (value == invalid){
        return invalid;
    }
    // outside of tracking window, also invalid
    else if (maxtracking != ReuseDistance::Infinity && value > maxtracking){
        return invalid;
    }
    // valid but not tracked individually
    else if (binindividual != ReuseDistance::Infinity && value > binindividual){
        return ShaveBitsPwr2(value);
    }
    // valid and tracked individually
    return value;
}

ReuseStats* ReuseDistance::GetStats(uint64_t id){
    return GetStats(id, false);
}

uint64_t ReuseStats::GetAccessCount(){
    return accesses;
}

uint64_t ReuseStats::GetMaximumDistance(){
    uint64_t max = 0;
    for (reuse_map_type<uint64_t, uint64_t>::const_iterator it = distcounts.begin(); it != distcounts.end(); it++){
        uint64_t d = it->first;
        if (d > max){
            max = d;
        }
    }
    return max;
}

void ReuseStats::Update(uint64_t dist){
    distcounts[GetBin(dist)] += 1;
    accesses++;
}

uint64_t ReuseStats::CountDistance(uint64_t d){
    if (distcounts.count(d) == 0){
        return 0;
    }
    return distcounts[d];
}

void ReuseStats::GetSortedDistances(vector<uint64_t>& dkeys){
    assert(dkeys.size() == 0 && "dkeys must be an empty vector");
    for (reuse_map_type<uint64_t, uint64_t>::const_iterator it = distcounts.begin(); it != distcounts.end(); it++){
        uint64_t d = it->first;
        dkeys.push_back(d);
    }
    sort(dkeys.begin(), dkeys.end());    
}

void ReuseStats::Print(ostream& f, bool annotate){
    vector<uint64_t> keys;
    GetSortedDistances(keys);

    if (annotate){
        ReuseStats::PrintFormat(f);
    }

    for (vector<uint64_t>::const_iterator it = keys.begin(); it != keys.end(); it++){
        uint64_t d = *it;
        if (d == invalid) continue;

        debug_assert(distcounts.count(d) > 0);
        uint32_t cnt = distcounts[d];

        debug_assert(cnt > 0);
        if (cnt > 0){
            uint64_t p = d / 2 + 1;
            if (binindividual == ReuseDistance::Infinity || d <= binindividual){
                p = d;
            }
            f << TAB
              << TAB << dec << p
              << TAB << d
              << TAB << cnt
              << ENDL;
        }
    }
}

void SpatialLocality::Init(uint64_t size, uint64_t bin, uint64_t max){
    //ReuseDistance::Init(size, ReuseDistance::Infinity);

    sequence = 1;
    capacity = size;
    binindividual = bin;
    maxtracking = max;

    assert(capacity > 0 && "window size must be a positive value");
    assert((maxtracking == INFINITY_REUSE || maxtracking >= binindividual) && "max tracking must be at least as large as individual binning");
}

ReuseStats* SpatialLocality::GetStats(uint64_t id, bool gen){
    ReuseStats* s = stats[id];
    if (s == NULL && gen){
        s = new ReuseStats(id, binindividual, maxtracking, SpatialLocality::Invalid);
        stats[id] = s;
    }
    return s;
}

void SpatialLocality::Process(ReuseEntry& r){
    uint64_t addr = r.address;
    uint64_t id = r.id;
    ReuseStats* stats = GetStats(id, true);
    debug_assert(stats);

    // find the address closest to addr
    uint64_t bestdiff = SpatialLocality::Invalid;

    if (awindow.size() > 0){

        map<uint64_t, uint64_t>::const_iterator it = awindow.upper_bound(addr);
        if (it == awindow.end()){
            it--;
        }

        // only need to check the values immediately equal, >, and < than addr
        for (uint32_t i = 0; i < 3; i++, it--){
            uint64_t cur = it->first;
            uint64_t seq = it->second;
            uint64_t diff = uint64abs(cur - addr);

            if (diff < bestdiff){
                bestdiff = diff;
            }

            if (it == awindow.begin()){
                break;
            }
        }
    }

    stats->Update(bestdiff);

    // remove the oldest address in the window
    if (swindow.size() > capacity){
        uint64_t a = swindow.front();
        swindow.pop_front();

        uint64_t v = awindow[a];
        if (v > 1){
            awindow[a] = v - 1;
        } else {
            awindow.erase(a);
        }
    }

    // insert the newest address into the window
    awindow[addr]++;
    swindow.push_back(addr);
}

void SpatialLocality::GetActiveAddresses(std::vector<uint64_t>& addrs){
    assert(addrs.size() == 0);

    for (map<uint64_t, uint64_t>::const_iterator it = awindow.begin(); it != awindow.end(); it++){
        uint64_t addr = it->first;
        addrs.push_back(addr);
    }
}

